### PR TITLE
Improve performance of the Resize 3c/3p and Resize 8UC1

### PR DESCRIPTION
**Pull request by anna-khakimova**
Performance report for 3C/3P, 4C/4P:
[Resize3_4C_NEON_RPI3_downscale_perf_improved.xlsx](https://github.com/openvinotoolkit/openvino/files/6295375/Resize3_4C_NEON_RPI3_downscale_perf_improved.xlsx)

Performance report:
[Resize1C_NEON_RPI3_downscale improve.xlsx](https://github.com/openvinotoolkit/openvino/files/6328371/Resize1C_NEON_RPI3_downscale.improve.xlsx)

### CODE REVIEW

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 1]
```diff
@@ -218,14 +218,15 @@ CV_ALWAYS_INLINE void horizontal_anyLPI(std::array<std::array<uint8_t*, 4>, chan
...
+                                      const short clone[], const uchar _mask_horizontal[],
+									  const int length) {
     constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
     const int half_nlanes = nlanes / 2;
     GAPI_Assert(length >= half_nlanes);
 
     const int shift = static_cast<int>(half_nlanes / 4);
 
-    uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14, 1, 5, 9, 13, 3, 7, 11, 15 };
+    //uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14, 1, 5, 9, 13, 3, 7, 11, 15 };
```
Comment (ilyachur) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:229]: Can we remove commented code?

**Review by ilyachur — COMMENTED** [review]
Do we have tests which show performance improvements?

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 1 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:229]: As you can see it's draft. I'll request a review as soon as this PR be ready for review.

**Review by ilya-lavrenov — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 2]
```diff
@@ -215,20 +215,120 @@ CV_ALWAYS_INLINE void horizontal_anyLPI(std::array<std::array<uint8_t*, 4>, chan
...
+    constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
+    const int half_nlanes = nlanes / 2;
+    GAPI_Assert(length >= half_nlanes);
+
+    v_int16 b0 = vx_setall_s16(beta[0]);
+    v_int16 b1 = vx_setall_s16(beta[1]);
+    v_int16 b2 = vx_setall_s16(beta[2]);
+    v_int16 b3 = vx_setall_s16(beta[3]);
+
+    //v_int16 lo, hi;
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:229]: Remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 3]
```diff
@@ -215,20 +215,120 @@ CV_ALWAYS_INLINE void horizontal_anyLPI(std::array<std::array<uint8_t*, 4>, chan
...
+            v_int16 t0 = v_mulhrs(v_sub_wrap(val0_0, val1_0), b0);
+            v_int16 t1 = v_mulhrs(v_sub_wrap(val0_1, val1_1), b1);
+            v_int16 t2 = v_mulhrs(v_sub_wrap(val0_2, val1_2), b2);
+            v_int16 t3 = v_mulhrs(v_sub_wrap(val0_3, val1_3), b3);
+
+            v_int16 r0 = v_add_wrap(val1_0, t0);
+            v_int16 r1 = v_add_wrap(val1_1, t1);
+            v_int16 r2 = v_add_wrap(val1_2, t2);
+            v_int16 r3 = v_add_wrap(val1_3, t3);
+#if 0
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:253]: Remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 4]
```diff
@@ -215,20 +215,120 @@ CV_ALWAYS_INLINE void horizontal_anyLPI(std::array<std::array<uint8_t*, 4>, chan
...
+                                      const short clone[],
+                                      const int length) {
     constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
     const int half_nlanes = nlanes / 2;
     GAPI_Assert(length >= half_nlanes);
 
     const int shift = static_cast<int>(half_nlanes / 4);
 
-    uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14, 1, 5, 9, 13, 3, 7, 11, 15 };
+    //uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14, 1, 5, 9, 13, 3, 7, 11, 15 };
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:327]: Remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 5]
```diff
@@ -296,77 +396,24 @@ CV_ALWAYS_INLINE void calcRowLinear_8UC_Impl_(std::array<std::array<uint8_t*, 4>
                                               const short    mapsx[],
                                               const short    beta[],
                                                   uint8_t    tmp[],
+                                            const uint8_t    _mask_horizontal[],
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:399]: alignment

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 6]
```diff
@@ -453,12 +500,13 @@ void calcRowLinear_8U(C3, std::array<std::array<uint8_t*, 4>, 3>& dst,
                       const short    mapsx[],
                       const short    beta[],
                           uint8_t    tmp[],
+                    const uint8_t    vmask[],
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:503]: alignment

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 7]
```diff
@@ -470,23 +518,24 @@ void calcRowLinear_8U(C4, std::array<std::array<uint8_t*, 4>, 4>& dst,
...
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:521]: alignment

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 8]
```diff
@@ -470,23 +518,24 @@ void calcRowLinear_8U(C4, std::array<std::array<uint8_t*, 4>, 4>& dst,
...
+                    const uint8_t    hmask[],
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:522]: alignment

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 9]
```diff
@@ -470,23 +518,24 @@ void calcRowLinear_8U(C4, std::array<std::array<uint8_t*, 4>, 4>& dst,
...
-                                      const uchar* tmp, const short mapsx[],
+                                      const uchar* tmp, const short mapsx[], const uchar _mask_horizontal[],
                                       const short clone[], const int length) {
     constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
     const int half_nlanes = nlanes / 2;
     GAPI_Assert(length >= half_nlanes);
 
-    uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14,
-                                      1, 5, 9, 13, 3, 7, 11, 15 };
+    //uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14,
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:537]: remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 10]
```diff
@@ -470,23 +518,24 @@ void calcRowLinear_8U(C4, std::array<std::array<uint8_t*, 4>, 4>& dst,
...
+    //                                  1, 5, 9, 13, 3, 7, 11, 15 };
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:538]: remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 11]
```diff
@@ -597,6 +646,7 @@ void calcRowLinear_8UC1(uint8_t* dst[],
 
         if (4 == lpi) {
             // vertical pass
+#if 0
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:649]: remove

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 12]
```diff
@@ -648,9 +698,14 @@ void calcRowLinear_8UC1(uint8_t* dst[],
...
             }
+#else
+            vertical_4LPI(src0, src1, tmp, beta, inSz.width);
+#endif
 
             // horizontal pass
-             horizontal_4LPI(dst, tmp, mapsx, clone, outSz.width);
+            uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14,
+                                                1, 5, 9, 13, 3, 7, 11, 15 };
+             horizontal_4LPI(dst, tmp, mapsx, _mask_horizontal, clone, outSz.width);
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:708]: alignment

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 13]
```diff
@@ -689,7 +744,9 @@ void calcRowLinear_8UC1(uint8_t* dst[],
             }
 
             // horizontal pass
-            horizontal_4LPI(dst, tmp, mapsx, clone, outSz.width);
+            uchar _mask_horizontal[nlanes] = { 0, 4, 8, 12, 2, 6, 10, 14,
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:747]: Move to scratch buffer

File: inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp [hunk 14]
```diff
@@ -885,6 +892,10 @@ struct linearScratchDesc {
                     outH * sizeof(alpha_t)     +
                     outH * sizeof(index_t) * 2 +
                      inW * sizeof(T) * lpi * chanNum;
+#ifdef HAVE_NEON
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp:895]: I would reinvent this, at least add comment

File: inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp [hunk 15]
```diff
@@ -914,11 +925,28 @@ static void initScratchLinear(const cv::GMatDesc& in,
...
 
     auto *alpha = scr.alpha;
     auto *clone = scr.clone;
     auto *index = scr.mapsx;
+#ifdef HAVE_NEON
+    auto* vmask = scr.shuffle_mask_vertical;
+    auto* hmask = scr.shuffle_mask_horizontal;
+
+    if (std::is_same<T, uint8_t>::value) {
+        for (int i = 0; i < 16/4; ++i) {
```
Comment (TolyaTalamanov) [inline inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp:938]: I would put a comment

**Review by TolyaTalamanov — APPROVED** [review]
LGTM :+1

Only cosmetic fixes required

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp [hunk 16]
```diff
@@ -215,20 +215,120 @@ CV_ALWAYS_INLINE void horizontal_anyLPI(std::array<std::array<uint8_t*, 4>, chan
     }
 }
 
+CV_ALWAYS_INLINE void vertical_4LPI(const uint8_t* src0[], const uint8_t* src1[],
+                                    uchar tmp[], const short beta[], const int length) {
+    constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
+    const int half_nlanes = nlanes / 2;
```
Comment (AsyaPronina) [inline inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:221]: why not `constexpr`?

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 3 above)
Comment (AsyaPronina) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp:253]: Is this branch really needed here?

**Review by AsyaPronina — APPROVED** [review]
approved with 2 minor comments:
- change `const` to `constexpr`
- remove `#if 0 blocks`

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 2 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 3 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Removed.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 4 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 6 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 7 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 8 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 9 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 10 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 11 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Removed.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 12 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Done.

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 13 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Reworked

File: inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp (hunk 14 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp]: Reworked

File: inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp (hunk 15 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp]: Reworked

File: inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp (hunk 16 above)
Comment (anna-khakimova) [reply inference-engine/src/preprocessing/arm_neon/ie_preprocess_gapi_kernels_neon.cpp]: Done.

**Comment by anna-khakimova** [comment]
> approved with 2 minor comments:
> 
> * change `const` to `constexpr`
> * remove `#if 0 blocks`

Done.

**Comment by anna-khakimova** [comment]
@ilya-lavrenov Could you please merge the PR?
One check haven't finished yet because it hung out on node allocation step.

**Comment by ilya-lavrenov** [comment]
I can merge only green CI - currently, it's still running..

**Comment by anna-khakimova** [comment]
@ilya-lavrenov As we all know, unfortunately, OpenVINO CI is very unstable entity. And if we always will waiting for when it become "green", we may wait for very very very long time. It's block the task finishing and work on other issues.

**Pull request merged**